src/arithmetic/canonical_simplify.cc

```diff
@@ -0,0 +1,884 @@
+    ret.CopyOnWrite()->AddToSelf(-op->value);
+  } else if (const auto* op = b.as<SumExprNode>()) {
+    ret.CopyOnWrite()->AddToSelf(GetRef<SumExpr>(op), -1);
+  } else {
+    ret.CopyOnWrite()->AddToSelf(ToSplitExpr(b), -1);
+  }
+  return ret;
+}
+
+
+Expr CanonicalSimplifier::Impl::
+Mutate_(const Mul* op, const Expr& self) {
+  if (!IsIndexType(op->type)) {
+    return Rewriter::Mutate_(op, self);
+  }
+  // normalize
+  Expr a = this->CanonicalMutate(op->a);
+  Expr b = this->CanonicalMutate(op->b);
+
+  // const folding
+  Expr const_res = TryConstFold<Mul>(a, b);
+  if (const_res.defined()) return const_res;
+
+  // x * c
+  if (a.as<IntImm>()) {
+    std::swap(a, b);
+  }
+  if (const auto* bconst = b.as<IntImm>()) {
+    if (a.as<SumExprNode>()) {
+      SumExpr ret(std::move(a.node_));
+      ret.CopyOnWrite()->MulToSelf(bconst->value);
+      return ret;
+    } else {
+      SplitExpr ret = ToSplitExpr(std::move(a));
+      ret.CopyOnWrite()->MulToSelf(bconst->value);
+      return ret;
+    }
+  }
+
+  // normal path.
+  a = Normalize(a);
+  b = Normalize(b);
+  if (op->a.same_as(a) && op->b.same_as(b)) {
+    return self;
+  } else {
+    return Mul::make(a, b);
+  }
+}
+
+
+bool CanonicalSimplifier::Impl::
+TryLinearEquation(const SumExprNode* psum,
+                  int64_t coeff,
+                  SumExpr* out_divisible,
+                  SumExpr* out_non_divisible) {
+  auto divisible = make_node<SumExprNode>();
+  auto non_divisible = make_node<SumExprNode>();
+  divisible->type = psum->type;
+  non_divisible->type = psum->type;
+
+  if (psum->base % coeff == 0) {
+    divisible->base = psum->base;
+  } else {
+    non_divisible->base = psum->base;
+  }
+  for (const auto& e : psum->args) {
+    if (e->scale % coeff == 0) {
+      divisible->args.push_back(e);
+    } else {
+      non_divisible->args.push_back(e);
+    }
+  }
+  *out_divisible = SumExpr(divisible);
+  *out_non_divisible = SumExpr(non_divisible);
+
+  if (non_divisible->base == 0 && non_divisible->args.size() == 0) {
+    return true;
+  }
+  if (parent_->CanProveGreaterEqual(divisible->Normalize(), 0) &&
+      parent_->CanProveGreaterEqual(non_divisible->Normalize(), 0)) {
+    return true;
+  } else {
+    return false;
+  }
+}
+
+SplitExpr CanonicalSimplifier::Impl::
+SplitDivConst(SplitExpr lhs, int64_t cval) {
+  if (lhs->scale % cval == 0) {
+    lhs.CopyOnWrite()->scale /= cval;
+    return lhs;
+  }
+
+  if (cval % lhs->scale == 0) {
+    int64_t scaled_cval = cval / lhs->scale;
+    if (lhs->upper_factor == SplitExprNode::kPosInf ||
+        lhs->upper_factor % (lhs->lower_factor * scaled_cval) == 0) {
+      // directly fold division.
+      lhs.CopyOnWrite()->scale = 1;
+      lhs.CopyOnWrite()->lower_factor *= scaled_cval;
+      lhs->Verify();
+      return lhs;
+    } else if (lhs->upper_factor <= (lhs->lower_factor * scaled_cval)) {
+      // (x % c1) / c2  => 0 when c2 >= c1
+      return ToSplitExpr(make_zero(lhs.type()));
```

Comment:
Can we also return zero when cval % lhs->scale != 0?  I mean the below looks correct in any cases.
```c
if (lhs->upper_factor <= (lhs->lower_factor * cval / lhs->scale))
   return ToSplitExpr(make_zero(lhs.type()));
```

Reply:
Because the mul and division are not necessarily exchangeable, in the case of cval % lhs->scale != 0, we will need to consider the consequence more carefully, because it is rare to have such case, we just skip the optimization